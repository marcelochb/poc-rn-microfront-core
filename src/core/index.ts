/* istanbul ignore file */
export * from './constants';
export * from './services';