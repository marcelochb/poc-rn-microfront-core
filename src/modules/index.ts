
/* istanbul ignore file */
export * from './loan';
