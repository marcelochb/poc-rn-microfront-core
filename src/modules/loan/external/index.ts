/* istanbul ignore file */
export * from './datasources'