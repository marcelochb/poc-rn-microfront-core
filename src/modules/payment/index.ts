/* istanbul ignore file */
export * from './domain/entities';
export {usePaymentListController} from './presenter/list/controller';
export {usePaymentDetailController} from './presenter/detail/controller';