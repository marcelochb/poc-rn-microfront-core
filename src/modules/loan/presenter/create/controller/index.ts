import { IControllerFormData, IControllerGetData } from "@poc/interfaces"
import { useFormik } from "formik"
import { useState } from "react"
import { CoreConstants } from "../../../../../core"
import { ICreateLoanUsecase, LoanEntity } from "../../../domain"
import { loanBind } from "../../loan_bind"

export const useLoanCreateController:IControllerFormData<LoanEntity> = (callback?: Function) => {
  const [error, setError] = useState(false);
  const useCase = loanBind.get<ICreateLoanUsecase>(CoreConstants.CreateLoanUsecase);
  const formik = useFormik<LoanEntity>({
    initialValues: new LoanEntity({amount:'',id: '',name: '', type: ''}),
    onSubmit: async (value) => {
      try {
        await useCase.call(value);
        callback?.call(this);
      } catch (error) {
        setError(true);
      }
    }
  });
  return {
    data: {
      amount: formik.values.amount,
      id: formik.values.id,
      name: formik.values.name,
      type: formik.values.type
    }, 
    loading: formik.isSubmitting, 
    error,
    onSubmit: formik.handleSubmit,
    onChange: formik.handleChange,
  }
}