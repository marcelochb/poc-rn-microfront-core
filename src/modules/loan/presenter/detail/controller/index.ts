import { IControllerGetData } from "@poc/interfaces";
import { useRoute } from "@react-navigation/native";
import React from "react";
import { container } from "tsyringe";
import { CoreConstants } from "../../../../../core";
import { IGetByLoanUsecase, LoanEntity } from "../../../domain";
import { loanDetailDependences } from "../bind";
import { ILoanDetailNavigationRoute } from "../interface";

loanDetailDependences();
export const useLoanDetailController:IControllerGetData<LoanEntity> = () => {
  const [loading, setLoading] = React.useState(true);
  const [error, setError] = React.useState(false);
  const [loan, setLoan] = React.useState<LoanEntity>({} as LoanEntity);
  const useCase = container.resolve<IGetByLoanUsecase>(CoreConstants.GetByLoanUsecase);
  const route = useRoute<ILoanDetailNavigationRoute>();
  React.useEffect(
    () => {
      const loadData = async () => {
        try {
          setError(false);
          setLoading(true);
          const response = await useCase.call({id: route.params.id});
          setLoan(response);
          setLoading(false);
        } catch (error) {
          setLoading(false);
          setError(true);
        }
      };
      loadData();
    },[]
  )

  return {loading, error, data:loan}
}