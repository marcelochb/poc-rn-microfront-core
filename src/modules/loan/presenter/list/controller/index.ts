import { useCallback, useEffect, useState } from "react"
import { container } from "tsyringe";
import { LoanConstants } from "../../../../../core";
import { IGetListLoanUsecase, LoanEntity } from "../../../domain";
import { loanListDependences } from "../bind";
import { IControllerGetData } from "@poc/interfaces";
import { useFocusEffect } from "@react-navigation/native";

loanListDependences();
export const useLoanListController = ():IControllerGetData<LoanEntity[]> => {
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(false);
  const [loans, setLoans] = useState<LoanEntity[]>([]);
  const useCase = container.resolve<IGetListLoanUsecase>(LoanConstants.GetListLoanUsecase);

  useFocusEffect(
    useCallback(() => {
      let isActive = true;
  
      const fetchUser = async () => {
        try {
          const response = await useCase.call();
          setLoans(response);
          setLoading(false);
        } catch (error) {
          setLoading(false);
          setError(true);
        }
      };
  
      if (isActive) fetchUser();
  
      return () => {
        isActive = false;
        setLoading(true);
      };
    }, [])
  );
  
  return {loading, error, data:loans}
}