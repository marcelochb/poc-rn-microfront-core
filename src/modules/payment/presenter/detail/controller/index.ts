import { IControllerGetData } from "@poc/interfaces";
import { useRoute } from "@react-navigation/native";
import React from "react";
import { container } from "tsyringe";
import { CoreConstants } from "../../../../../core";
import { IGetByPaymentUsecase, PaymentEntity } from "../../../domain";
import { IPaymentDetailNavigationRoute } from "../interface";

export const usePaymentDetailController:IControllerGetData<PaymentEntity> = () => {
  const [loading, setLoading] = React.useState(true);
  const [error, setError] = React.useState(false);
  const [payment, setPayment] = React.useState<PaymentEntity>({} as PaymentEntity);
  const useCase = container.resolve<IGetByPaymentUsecase>(CoreConstants.GetByPaymentUsecase);
  const route = useRoute<IPaymentDetailNavigationRoute>();
  React.useEffect(
    () => {
      const loadData = async () => {
        try {
          setError(false);
          setLoading(true);
          const response = await useCase.call({id: route.params.id});
          setPayment(response);
          setLoading(false);
        } catch (error) {
          setLoading(false);
          setError(true);
        }
      };
      loadData();
    },[]
  )

  return {loading, error, data:payment}
}