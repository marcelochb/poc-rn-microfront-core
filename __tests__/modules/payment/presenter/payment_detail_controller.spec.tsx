import 'reflect-metadata';
import { container } from 'tsyringe';
import { CoreConstants } from '../../../../src/core';
import { useLoanDetailController } from '../../../../src/modules';
import { Text } from 'react-native';
import { render, screen, waitForElementToBeRemoved } from '@testing-library/react-native';
import React from 'react';
import { paymentEntity } from '../../../../src/modules/payment/external';
import { usePaymentDetailController } from '../../../../src/modules/payment/presenter/detail/controller';

jest.mock("@react-navigation/native", () => ({
  ...jest.requireActual("@react-navigation/native"),
  useRoute: jest.fn().mockImplementation(() => {return {params: {id:'1'}}} ),
}));

interface IProps {
  id: string;
}

describe('Payment Deail Controller =>',() => {
  it('When success request should retun IContrllerData<PaymentEntity>', async () => {
    class GetByPaymentMockUsecase {
      call = jest.fn().mockImplementation( async ({id}:IProps) => Promise.resolve(paymentEntity));
    }
    container.register(CoreConstants.GetByPaymentUsecase,{useValue: new GetByPaymentMockUsecase()});

    const MyComponent = () => {
      const {data, loading} = usePaymentDetailController();
      return (
        <Text>{loading ? 'loading':data.recipient}</Text>
        );
    }
    render(
        <MyComponent />
    )
    await waitForElementToBeRemoved(() => screen.getByText('loading'));
    expect(screen.toJSON()).toMatchSnapshot();        
    
  })
  it('When fail request should retun erro=true', async () => {
    class GetByPaymentMockUsecase {
      call = jest.fn().mockRejectedValue(async({id}:IProps) =>{throw Error});
    }
    container.register(CoreConstants.GetByPaymentUsecase,{useValue: new GetByPaymentMockUsecase()});

    const MyComponent = () => {
      const {error,loading} = usePaymentDetailController();
      return (
        <Text>{loading 
                    ? 'loading' 
                    : error ? 
                        'comunication fail'
                        : 'sucesso'
              }
        </Text>
      );
    }
    render(
        <MyComponent />
    )
    await waitForElementToBeRemoved(() => screen.getByText('loading'));
    expect(screen.toJSON()).toMatchSnapshot();        
  })
})