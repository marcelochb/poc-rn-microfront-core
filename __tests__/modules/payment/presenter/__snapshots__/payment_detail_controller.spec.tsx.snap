// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Payment Deail Controller => When fail request should retun erro=true 1`] = `
<Text>
  comunication fail
</Text>
`;

exports[`Payment Deail Controller => When success request should retun IContrllerData<PaymentEntity> 1`] = `
<Text>
  teste
</Text>
`;
